### PR TITLE
restructure; reorganizing the code to make validation rules simpler

### PR DESCRIPTION
**Pull request by mangopeaches**
Hey there!

I was trolling around for hacktoberfest and happened upon your project. I saw your initial issues and wanted to contribute, but when I got started I started changing some structure and started having fun with it. You by no means have to accept this PR, I just thought I'd give my two cents on a structure that serves the same end goal, but reduces the code a bit and makes it a little more flexible and user friendly. Again, this is just my opinion and I mean no offense, just trying to help!

I restructured some things such that it behaves in the following way:
There is one Validator class. This class takes a single Type or an array or Types and performs validation on each value passed in. It also has an errors() method which will return bool indicating if all the Types passed validation, and has a errors() member function, which will return the array of errors if any Types failed validation.

I also added a single Type parent abstract class which implements the majority of the functionality the Type classes need. Then each individual Type (Email, Numeric, Name, etc) need only implement a single isValid() method, which will perform the actual validation for the given type.

I made this php 5.6 friendly (sorry, I'm still stuck on this version and I know a lot of others are as well). I added a phpunit.xml file such that you can execute tests with just `phpunit` and implemented all the tests for the new structure, all of which should be working.

Let me know if you have any questions, and again, I'm not trying to impose or change your project, just thought I'd share my thoughts; you should not feel obliged to merge this if you prefer your initial structure.

Thanks,
Tom

### CODE REVIEW

File: phpunit.xml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (2DSharp) [inline phpunit.xml]: Please remove this file from the commit.

Comment (mangopeaches) [reply phpunit.xml]: done!

**Review by 2DSharp — COMMENTED** [review]
Hi, thanks for your PR. It seems like you are trying to follow a different approach than what this project currently follows. One of the primary goals of this project is to provide Value Objects to be used and reused in other projects and *not* to be just a validation library. The idea of one god validation class might work in some situations, but the ideology of this project believes otherwise.

The way the VOs are supposed to work is somewhat in this manner:

- The user enters some data, for example, an email address. Gets stored in `$_POST['email]`.
- We make our services on the domain layer require the "Type" `Email` as a field as it corresponds to a field in the database.
- This type checking is done by the validator specified while calling the VO. If no validator is specified, the default validator for that type will be used.
- If the validation fails, you get an `InvalidArgumentException` just like you'd expect to get an exception when you try to put in string instead of an integer in a function that requires an integer.

As you can see, that it allows for some flexibility to the developer. The developer can create their own types and/or implement their own validators for some specific types they wish to. Everything is loosely coupled and it becomes easy to switch parts easily.

File: src/Type/Numeric.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<?php
+namespace GreenTea\Phypes\Type;
+
+use GreenTea\Phypes\Type\Type;
+
+/**
+ * Numeric class validates supplied input is indeed a numeric value.
+ * @author Mangopeaches <https://github.com/mangopeaches>
+ */
+class Numeric extends Type
```
Comment (2DSharp) [inline src/Type/Numeric.php:10]: The idea of a numeric type seems nice, maybe you could create a different version that complies with the design of the project?

**Comment by mangopeaches** [comment]
Hey,

Thanks for the explanation. I definitely got what you were going for, I was just having a little fun in the spirit of Hacktoberfest and this was a different approach at a similar problem. I will keep my fork and run with that as a separate project (thanks for giving me the opportunity to get started with a validation library I've been meaning to build :) ) I honestly didn't expect you to merge that, as I knew it was different from your approach and you have a different vision, but never hurts to share. Feel free to close out this PR, no worries at all, I had fun doing it! I will definitely give you a PR that conforms to what you have going on and add Numeric and potentially some other rules if you've got any other specific ones in mind. Feel free to create some additional issues for other types and I'll knock some out when I get the time.

Thanks!
Tom

**Comment by 2DSharp** [comment]
Great! Your approach should work in a different use scenario and I would like to see how that proceeds. Some PRs for some other types would be a great help, and the Hacktoberfest T-shirts should be knocking at your door. Couldn't get to the PRs sooner for some reason, I would be able to actively review them now. 

Thanks for being a good sport.

**Pull request closed without merging**